private/protocol/json/encode.go

```diff
@@ -0,0 +1,299 @@
+func (e *listEncoder) Map() protocol.MapEncoder {
+	temp := e.encoder
+	temp.scope = newScope(e.buf, temp.scope)
+	return &mapEncoder{temp, ""}
+}
+
+// List return a new listEncoder while creating a new scope for the encoder.
+func (e *listEncoder) List() protocol.ListEncoder {
+	temp := e.encoder
+	temp.scope = newScope(e.buf, temp.scope)
+	return &listEncoder{temp, ""}
+}
+
+// Start will open a new scope for a list and write the given key.
+func (e *listEncoder) Start() {
+	e.encoder.scope.parent.writeSep()
+	e.writeKey(e.k)
+	e.WriteListStart()
+}
+
+// End will close the list.
+func (e *listEncoder) End() {
+	e.WriteListEnd()
+}
+
+// ListAddValue will add the value to the list.
+func (e *listEncoder) ListAddValue(v protocol.ValueMarshaler) {
+	e.encoder.scope.writeSep()
+	e.writeValue(v)
+}
+
+// ListAddFields will set the nested type's fields to the list.
+func (e *listEncoder) ListAddFields(m protocol.FieldMarshaler) {
+	e.encoder.scope.writeSep()
+	e.writeObject(func(enc encoder) error {
+		enc.scope = newScope(enc.buf, enc.scope)
+		nested := Encoder{encoder: enc}
+		m.MarshalFields(&nested)
+		return nested.err
+	})
+}
+
+// A mapEncoder encodes key values pair map values for the JSON encoder.
+type mapEncoder struct {
+	encoder encoder
+	k       string
+}
+
+// Start will open a new scope for a list and write the given key.
+func (e *mapEncoder) Start() {
+	e.encoder.scope.parent.writeSep()
+	e.encoder.writeKey(e.k)
+	e.encoder.WriteMapStart()
+}
+
+// End will close the list.
+func (e *mapEncoder) End() {
+	e.encoder.WriteMapEnd()
+}
+
+// Map will create a new scope and return a mapEncoder.
+func (e *mapEncoder) Map(k string) protocol.MapEncoder {
+	temp := e.encoder
+	temp.scope = newScope(e.encoder.buf, temp.scope)
+	return &mapEncoder{temp, k}
+}
+
+// List will create a new scope and return a listEncoder
+func (e *mapEncoder) List(k string) protocol.ListEncoder {
+	temp := e.encoder
+	temp.scope = newScope(e.encoder.buf, temp.scope)
+	return &listEncoder{temp, k}
+}
+
+// MapSetValue sets a map value.
+func (e *mapEncoder) MapSetValue(k string, v protocol.ValueMarshaler) {
+	e.encoder.scope.writeSep()
+	e.encoder.writeKey(k)
+	e.encoder.writeValue(v)
+}
+
+// MapSetFields will set the nested type's fields under the map.
+func (e *mapEncoder) MapSetFields(k string, m protocol.FieldMarshaler) {
+	e.encoder.scope.writeSep()
+	e.encoder.writeKey(k)
+	e.encoder.writeObject(func(enc encoder) error {
+		enc.scope = newScope(enc.buf, enc.scope)
+		nested := Encoder{encoder: enc}
+		m.MarshalFields(&nested)
+		return nested.err
+	})
+}
+
+type encoder struct {
+	buf      *bytes.Buffer
+	fieldBuf *protocol.FieldBuffer
+	scope    *scope
+	err      error
+}
+
+func (e encoder) encode() ([]byte, error) {
+	if e.err != nil {
+		return nil, e.err
+	}
+
+	// Close the root object
+	e.buf.WriteByte('}')
+
+	return e.buf.Bytes(), nil
+}
+
+func (e *encoder) writeKey(k string) {
+	e.buf.WriteByte('"')
+	e.buf.WriteString(k) // TODO escape?
```

Comment:
todo still needed? I don't think the key is supposed to be escaped, but not positive.